### PR TITLE
HDWallet tests and code cleanup

### PR DESCRIPTION
**Pull request by catenocrypt**
## Description

Added new/more HDWallet tests, some code cleanup: remove some unreachable branches, add error checks.

## Testing instructions

Standard unit tests.

## Types of changes

* Code cleanup, New tests

## Checklist

- [x] Prefix PR title with `[WIP]` if necessary.
- [x] Add tests to cover changes as needed.
- [x] Update documentation as needed.
- [x] If there is a related Issue, mention it in the description (e.g. Fixes #<issue_number> ).

### CODE REVIEW

File: src/HDWallet.cpp [hunk 1]
```diff
@@ -148,29 +149,22 @@ std::optional<PublicKey> HDWallet::getPublicKeyFromExtended(const std::string& e
     if (!deserialize(extended, curve, hasher, &node)) {
         return {};
     }
+    if (node.curve->params == nullptr)
+    {
```
Comment (hewigovens) [inline src/HDWallet.cpp:153]:
```suggestion
    if (node.curve->params == nullptr) {
```

File: tests/HDWalletTests.cpp [hunk 2]
```diff
@@ -21,13 +21,42 @@ namespace TW {
 TEST(HDWallet, privateKeyFromXPRV) {
     const std::string xprv = "xprv9yqEgpMG2KCjvotCxaiMkzmKJpDXz2xZi3yUe4XsURvo9DUbPySW1qRbdeDLiSxZt88hESHUhm2AAe2EqfWM9ucdQzH3xv1HoKoLDqHMK9n";
     auto privateKey = HDWallet::getPrivateKeyFromExtended(xprv, DerivationPath(TWPurposeBIP44, TWCoinTypeBitcoinCash, 0, 0, 3));
+    ASSERT_TRUE(privateKey);
```
Comment (hewigovens) [inline tests/HDWalletTests.cpp:24]: should we use `ASSERT_NE`?

File: tests/interface/TWHDWalletTests.cpp [hunk 3]
```diff
@@ -307,6 +327,35 @@ TEST(HDWallet, PublicKeyFromZ) {
     assertStringsEqual(address4, "bc1qm97vqzgj934vnaq9s53ynkyf9dgr05rargr04n");
 }
 
+TEST(HDWallet, PublicKeyFromExtended_NIST256p1) {
+    const auto xpub = STRING("xpub6BosfCnifzxcFwrSzQiqu2DBVTshkCXacvNsWGYJVVhhawA7d4R5WSWGFNbi8Aw6ZRc1brxMyWMzG3DSSSSoekkudhUd9yLb6qx39T9nMdj");
+    const auto xpubAddr = WRAP(TWPublicKey, TWHDWalletGetPublicKeyFromExtended(xpub.get(), STRING("m/44'/888'/0'/0/0").get())); // Neo
+    ASSERT_TRUE(xpubAddr.get() != nullptr);
```
Comment (hewigovens) [inline tests/interface/TWHDWalletTests.cpp:333]: ASSERT_NE(xpubAddr.get(), nullptr)

File: tests/HDWalletTests.cpp (hunk 2 above)
Comment (catenocrypt) [reply tests/HDWalletTests.cpp:24]: return value here is optional, boolean true mean it has value.

File: tests/interface/TWHDWalletTests.cpp (hunk 3 above)
Comment (catenocrypt) [reply tests/interface/TWHDWalletTests.cpp]: Thnx, this better.

File: src/HDWallet.cpp [hunk 4]
```diff
@@ -287,15 +283,18 @@ HDNode getMasterNode(const HDWallet& wallet, TWCurve curve) {
 const char* curveName(TWCurve curve) {
     switch (curve) {
     case TWCurveSECP256k1:
+    default:
```
Comment (vikmeup) [inline src/HDWallet.cpp:286]: should we crash on default case? instead of returning `SECP256K1_NAME`

Comment (catenocrypt) [reply src/HDWallet.cpp:286]: Not sure what's best.  I'm reluctant to add unreachable/untestable lines of code; all 6 enum values are handled, so it's not possible to drive into separate default line.  Now I will delete the default line, though that may trigger compiler warning.

Comment (catenocrypt) [reply src/HDWallet.cpp]: Solved by introducing TWCurveNone value.

File: src/PrivateKey.cpp [hunk 5]
```diff
@@ -162,17 +164,20 @@ Data PrivateKey::sign(const Data& digest, TWCurve curve) const {
...
         result.resize(64);
         const auto publicKey = getPublicKey(TWPublicKeyTypeED25519Blake2b);
         ed25519_sign_blake2b(digest.data(), digest.size(), bytes.data(),
                              publicKey.bytes.data(), result.data());
+        success = true;
     } break;
     case TWCurveED25519Extended: {
         result.resize(64);
         const auto publicKey = getPublicKey(TWPublicKeyTypeED25519Extended);
         ed25519_sign_ext(digest.data(), digest.size(), bytes.data(), extensionBytes.data(), publicKey.bytes.data(), result.data());
```
Comment (vikmeup) [inline src/PrivateKey.cpp:179]: can we follow same format?

here is another variation:
```
success = ecdsa_sign_digest(&nist256p1, bytes.data(), digest.data(), result.data(),
                                    result.data() + 64, nullptr) == 0;
```

File: src/PrivateKey.cpp [hunk 6]
```diff
@@ -162,17 +164,20 @@ Data PrivateKey::sign(const Data& digest, TWCurve curve) const {
...
+        success = true;
```
Comment (vikmeup) [inline src/PrivateKey.cpp:180]: does `ed25519_sign_ext` function success by itself? maybe we should check that

Comment (catenocrypt) [reply src/PrivateKey.cpp:180]: That was my first thought as well, but some trezor-crypto sign methods have bool return value, some have none.  I guess that's why the original optimistic success = true initialization.  It is safer to start out with false.

void ed25519_sign_ext(const unsigned char *m, size_t mlen, const ed25519_secret_key sk, const ed25519_secret_key skext, const ed25519_public_key pk, ed25519_signature RS);

**Pull request merged**